# PlasmidPredictor
Given some raw uncorrected long reads, such as those from PacBio or Oxford Nanopore, predict which plasmid should be present.  Assemblies of long read data can often miss out on plasmids, particularly if they are very small or have a copy number which is too high/low when compared to the chromosome. This software gives you an indication of which plasmids to expect, flagging potential issues with an assembly.

[![Build Status](https://travis-ci.org/andrewjpage/plasmidpredictor.svg?branch=master)](https://travis-ci.org/andrewjpage/plasmidpredictor)

# Installation
The only dependancy is Python3. Assuming you have python 3.3+ and pip installed, just run:
```
pip3 install git+git://github.com/andrewjpage/plasmidpredictor.git
```

## Debian/Ubuntu (Trusty/Xenial)
To install Python3 on Ubuntu, as root run:
```
apt-get update -qq
apt-get install -y git python3 python3-setuptools python3-biopython python3-pip
pip3 install git+git://github.com/andrewjpage/plasmidpredictor.git
```

## Windows
Like virtually all Bioinformatics software, this software is unlikely to work on Windows. Try using a Linux virtual machine.

# Usage
## plasmidpredictor_database_downloader script
First of all you need plasmid database. There is a snapshot bundled with this repository for your convenience, or alternatively you can use the downloader script to get the latest data. You will need internet access for this step.

```
usage: plasmidpredictor_database_downloader [options]

Download PlasmidFinder database

optional arguments:
  -h, --help            show this help message and exit
  --output_prefix OUTPUT_PREFIX, -o OUTPUT_PREFIX
                        Output directory (default: plasmid_files)
  --verbose, -v         Turn on debugging (default: False)
  --version             show program's version number and exit
```

Just run:
```
plasmidpredictor_database_downloader 
```
You will now have a file called 'plasmid_files.fa' which can be used with the main script.

## plasmidpredictor script
This is the main script of the application. The mandatory inputs are the plasmid sequence FASTA file from the previous step, and a FASTQ file of long reads, which can be optionally gzipped.
```
usage: plasmidpredictor [options] plasmid_data.fa input.fastq

plasmid incompatibility group prediction from uncorrected long reads

positional arguments:
  plasmid_data          FASTA file containing plasmid data from downloader
                        script
  input_fastq           Input FASTQ file (optionally gzipped)

optional arguments:
  -h, --help            show this help message and exit
  --filtered_reads_file FILTERED_READS_FILE, -f FILTERED_READS_FILE
                        Filename to save matching reads to (default: None)
  --output_file OUTPUT_FILE, -o OUTPUT_FILE
                        Output file [STDOUT] (default: None)
  --max_gap MAX_GAP     Maximum gap for blocks to be contigous, measured in
                        multiples of the k-mer size (default: 3)
  --margin MARGIN       Flanking region around a block to use for mapping
                        (default: 10)
  --min_block_size MIN_BLOCK_SIZE
                        Minimum block size in bases (default: 130)
  --min_fasta_hits MIN_FASTA_HITS, -m MIN_FASTA_HITS
                        Minimum No. of kmers matching a read (default: 10)
  --min_perc_coverage MIN_PERC_COVERAGE, -c MIN_PERC_COVERAGE
                        Minimum percentage coverage of typing sequence to
                        report (default: 85)
  --min_kmers_for_onex_pass MIN_KMERS_FOR_ONEX_PASS
                        Minimum No. of kmers matching a read in 1st pass
                        (default: 10)
  --print_interval PRINT_INTERVAL, -p PRINT_INTERVAL
                        Print results every this number of reads (default:
                        1000)
  --kmer KMER, -k KMER  k-mer size (default: 13)
  --verbose, -v         Turn on debugging [False]
  --version             show program's version number and exit
```

### Required
__plasmid_data__: This is a FASTA file containing all of the plasmid typing sequencings. This is generated by the plasmidpredictor_database_downloader script. It comes from the plasmidFinder website, so please be sure to cite their paper (citation gets printed every time you run the script).

__input_fastq__: This is a single FASTQ file. It can be optionally gzipped. Alternatively input can be read from stdin by using the dash character (-) as the input file name.

### Options
__kmer__:  The most important parameter. Long reads have a high error rate, so if you set this too high, nothing will match (because it will contain errors). If you set it too low, everything will match, which isnt much use to you. Thinking about your data, on average how long of a stretch of bases can you get in your read without errors? This is what you should set your kmer to. For example, if you have an average of 1 error every 10 bases, then the ideal kmer would be 9.

__min_fasta_hits__: This is the minimum number of matching kmers in a read, for the read to be considered for analysis. It is a hard minimum threshold which is really there to speed things up. If you set this too high, then nothing will be returned.

__filtered_reads_file__: If you provide a filename for this option, all of the reads which are estimated to match one of the sequences are saved to a file. Only the region predicted to contain the sequence is saved. This can be used for downstream analysis, such as de novo assembly. This file should not already exist. 

__output_file__: By default the predicted sequence types are printed to screen (STDOUT). If a filename is provided, the predicted sequence types are instead printed to this file.  This file should not already exist. 

__print_interval__: Print out the predicted sequence type every X number of reads. This is where you are performing analysis in real time and want a quick result.

# Resource usage
For an 800Mbyte FASTQ file (unzipped) of long reads from a Oxford Nanopore MinION containing Salmonella required 80MB of RAM and took under 1 minute.


